src/style/Color.js

```diff
@@ -899,8 +899,11 @@ var Color = Base.extend(new function() {
                         + components.join(',') + ')';
         },
 
-        toCanvasStyle: function(ctx, matrix) {
-            if (this._canvasStyle)
+        toCanvasStyle: function(ctx, matrix, strokeMatrix) {
+            // strokeMatrix can change without triggering _changed here, so we
```

Comment:
Is this fixable by correctly triggering the _changed event when strokeMatrix changes?
It seems brittle to keep track of exceptions around caching

Reply:
I could manually trigger `_changed` in [this block](https://github.com/LLK/paper.js/blob/498097172d7925ced5bb13dcf913f4157ee0beb8/src/item/Item.js#L4446-L4455) but that seems more fragile to me-- the color code would no longer be responsible for ensuring its own caching behavior.

I could also just stick `if (strokeMatrix) this._changed();` at the top of this function.

Reply:
And come to think of it that would suffer from the same problem-- if `strokeMatrix` stopped being passed in, `_changed` would stop being called and the gradient would retain its previous incorrect transform

Reply:
I'm working on understanding the paper.js code, and it looks like the Color class already doesn't keep track of its own matrix; it's relying on the parent item to call _transform on the color any time the item's matrix changes, otherwise if you call toCanvasStyle twice in a row with different matrices you'll get the wrong result.

What if the color stored what was in the matrix and strokeMatrix, and used that to determine whether its cached style was still valid?

Are we assuming that any time a strokeMatrix is passed in, it implies that we're coloring a stroke?

Reply:
`strokeMatrix`, confusingly, actually refers to (IIRC) the matrix by which path points are transformed to obtain the view transformation. It's applied to the fill *and* stroke when `strokeScaling` is `false`.

The problem is that it would also need to be invalidated whenever the view transform changes.